### ISSUE TITLE
Attributes no longer indexing

### ISSUE THREAD

**Issue by hi-im-si** [WebDevStudios/wp-search-with-algolia#177]
**Describe the bug**
Upon following the guide for adding [custom attributes](https://github.com/WebDevStudios/wp-search-with-algolia/wiki/Custom-Attributes) I was able to successfully create multiple attributes to my indexed records but without warning, it has now unexpectedly stopped pushing new attributes to my records.

Here is a snippet of my algolia.php file which indexes various attributes into my records

```
function osc_algolia_wc_post_shared_attributes( array $shared_attributes, WP_Post $post) {
	$product = wc_get_product( $post );
	// Extract prices.
	$variations_count = 0;
	if ( $product instanceof WC_Product_Variable ) {
		$price = $product->get_variation_price( 'min', true );
		$regular_price = $product->get_variation_regular_price( 'min', true );
		$sale_price = $product->get_variation_sale_price( 'min', true );
		$max_price = $product->get_variation_price( 'max', true );
		$variations_count = count( $product->get_available_variations() );
                 $finance_price = get_post_meta($product->get_id(), 'lowest_regular_finance_payment', true);
	} else {
		$price = $max_price = $product->get_display_price();
		$regular_price = $product->get_display_price( $product->get_regular_price() );
		$sale_price = $product->get_display_price( $product->get_sale_price() );
	}

	$shared_attributes['product_type'] = (string) $product->get_type();
	$shared_attributes['price'] = (float) $price;
	$shared_attributes['regular_price'] = (float) $regular_price;
	$shared_attributes['sale_price'] = (float) $sale_price;
	$shared_attributes['finance_price'] = (float) $finance_price;
	$shared_attributes['is_on_sale'] = (bool) $product->is_on_sale();
	$shared_attributes['course_code'] = get_post_meta($product->get_id(), 'code', true);
        $shared_attributes['product_description'] = get_field('introduction', $post->ID);
        $shared_attributes['page_description'] = get_field('page_summary', $post->ID);

        return $shared_attributes;
}
add_filter( 'algolia_post_product_shared_attributes', 'osc_algolia_wc_post_shared_attributes', 10, 2 );
add_filter( 'algolia_searchable_post_product_shared_attributes', 'osc_algolia_wc_post_shared_attributes', 10, 2 );


function osc_algolia_post_page_shared_attributes( array $shared_attributes, WP_Post $post) {
    $shared_attributes['page_description'] = get_field('page_summary', $post->ID);
    $shared_attributes['page_content'] = get_field('content', $post->ID);

    return $shared_attributes;
}
add_filter( 'algolia_post_shared_attributes', 'osc_algolia_post_page_shared_attributes', 10, 2 );
add_filter( 'algolia_post_page_shared_attributes', 'osc_algolia_post_page_shared_attributes', 10, 2 );
add_filter( 'algolia_searchable_post_shared_attributes', 'osc_algolia_post_page_shared_attributes', 10, 2 );

function osc_posts_index_settings(array $settings)
{

    // Make Algolia search into the 'product_description' field when searching for results.
    // Using ordered instead of unordered would make words matching in the beginning of the attribute
    // make the record score higher.
    // @see https://www.algolia.com/doc/api-client/ruby/parameters#attributestoindex
    $settings['attributesToIndex'][] = 'unordered(product_description)';
    $settings['attributesToIndex'][] = 'unordered(page_description)';
    $settings['attributesToIndex'][] = 'unordered(page_content)';
    $settings['attributesToIndex'][] = 'unordered(course_code)';

    // Make Algolia return a pre-computed snippet of 100 chars as part of the result set.
    // @see https://www.algolia.com/doc/api-client/ruby/parameters#attributestohighlight
    $settings['attributesToSnippet'][] = 'product_description:100';
    $settings['attributesToSnippet'][] = 'page_description:100';
    $settings['attributesToSnippet'][] = 'page_content:100';
    $settings['attributesToSnippet'][] = 'course_code:100';

    // Always return the value we are filtering.
    return $settings;
}
add_filter('algolia_posts_product_index_settings', 'osc_posts_index_settings');
add_filter('algolia_posts_index_settings', 'osc_posts_index_settings');
add_filter('algolia_posts_page_index_settings', 'osc_posts_index_settings');
```

The attribute I am having trouble pushing into the index is:
`$shared_attributes['course_code'] = get_post_meta($product->get_id(), 'code', true);`
but I have also attempted to update some of the existing attributes for example adding `0.5` to the `price` just to debug but unfortunately does not do anything.

I am of course ensuring that once I update the PHP file, I am going into the backend to reindex the indices on All Posts/Products etc.

**Expected behavior**
Any custom attributes to be indexed correctly

**Additional context**
I can appreciate the code above may be a little messy but this is due to constant iteration in an attempt to get the sync working as expected.

**Comment by tw2113** [comment]
Have you done any core plugin updates between the last time it worked and now? Just trying to rule out that as possible factor.

Are you sure that that code meta line is returning anything? Possible it's empty value(s) somehow.

**Comment by hi-im-si** [comment]
Thanks for your reply @tw2113 - I can confirm that WordPress was upgraded in a recent deployment from 5.7 to 5.8

The code meta line is used in multiple other places throughout the site so the exact code was used in this instance. I am not seeing any errors in my logs but could I ask if there is a recommended way of debugging the `$shared_attributes` variable inside my `osc_algolia_wc_post_shared_attributes` function? my experience of debugging functions being assigned to filters is non-existent unfortunately. Thanks for your support thus far

**Comment by hi-im-si** [comment]
as an additional piece of information, this issue is only occurring on my Staging/Production sites. My local version pushes the attribute to the algolia index successfully.....which now creates even more confusion for me

**Comment by tw2113** [comment]
Typically my debugging tends to be either utilizing XDebug with PHPStorm to pause execution, including inside callbacks on filters, and observing what happens there at that time.

Alternatively, I've also emailed myself some `print_r` results before when I haven't been able to make use of more advanced debugging tools.

At this point, I'd probably check the algolia instance and a specific index where you're not seeing the code attribute/meta value, and then confirm via product if the WP side of things has a value for it. Visual cross checking just to be certain that there's data that should be getting included.

**Comment by hi-im-si** [comment]
Hey @tw2113 the visual cross-checking was done first and foremost but the most interesting part to this issue which I was hoping someone may be able to point me in a direction is that the following occurs:

- Staging/Production both DO NOT currently push new attributes AND after I deleted there indexes from the algolia backend, upon reindexing again all my records are successfully indexes but the aforementioned Attribute (course_code) still does not get pushed. Would this be a caching mechanism somewhere?

- the interesting part to this is spinning up my local dev environment and indexing locally, this elusive attribute now appears in my development index in algolia.

The common elements here are database are near identical albeit a couple of blog posts and products, codebase is obviously identical, Wordpress core and plugins are the same version numbers.

And so as you can appreciate this is quite confusing as to why only on staging and production that this new attribute fails to sync.

I can appreciate its not an easy one to provide a solution but if there are any parts of the plug-in I can temporarily edit to see what gets sent to algolia then any help would truly save my sanity 😅

Thanks again for your responses thus far 👍🏻

**Comment by mtozlu** [comment]
It may be because of a parameter replacement in Algolia API.
Please see this PR to fix the issue: #179

**Comment by tw2113** [comment]
> Staging/Production both DO NOT currently push new attributes AND after I deleted there indexes from the algolia backend, upon reindexing again all my records are successfully indexes but the aforementioned Attribute (course_code) still does not get pushed. Would this be a caching mechanism somewhere?

Anything caching related would be within the site in question itself, nothing on the plugin side.

Unsure if the comments/PR from @mtozlu are going to help here, but I'm hesitant to recommend outright replacing those index values as we do try to retain backwards compatibility where possible. Maybe add those as indexes instead of replacing. @richaber may have more insight about what's going on with that part too.

**Comment by mtozlu** [comment]
Yeah, I'm not sure either. Just took my slim chance in case it helps.

**Comment by hi-im-si** [comment]
Thanks again.
I have managed to get someway with this issue and it is still providing me frustration....1 step forward, 2 steps back.

I resorted to using the WP-CLI command which successfully synced my custom attribute 🎉 BUT another issue was introduced which I have created here: https://github.com/WebDevStudios/wp-search-with-algolia/issues/181 

I since tried to hit the reindex button in the CMS to see if the records remain with the new attribute but it unfortunately overwrites the records and my attribute is lost and so the question now is.......what is the difference in running the CLI command versus the reindex button in the WordPress CMS.
